### PR TITLE
julius: darwin build, add maintainer

### PR DESCRIPTION
**Pull request by matteo-pacini**
## Description of changes

Added Darwin support to Julius.

List of changes:
- Disabled `fixup_bundle` CMake directive - gave a few errors and doesn't really makes sense on Nix
- Disabled unit tests - they don't compile 
- Set min deployment target to macOS 11.0
- Cmake app bundle generation :
    - replaced `sips` with `convert` (`imagemagick`)
    - replaced `iconutil` with `png2icns` (`libicns`)
    - replaced `codesign` binary with the one from `darwin.sigtool`
- Added myself as second maintainer for Darwin

The screenshot below is from the compiled `julius.app` running on my M1 machine (macOS 14.5).

Screenshot 2024-06-07 at 17 52 43

Binaries tested both on `aarch64-darwin` and `aarch64-linux`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/games/julius/default.nix [hunk 1]
```diff
@@ -18,16 +25,42 @@ stdenv.mkDerivation rec {
...
-  buildInputs = [ SDL2 SDL2_mixer libpng ];
+  patches = [
+    # This disables the broken fixup_bundle and
+    # also disables unit tests.
+    ./darwin-fixes.patch
+  ];
+
+  postPatch = lib.optionalString stdenv.isDarwin ''
+    substituteInPlace src/platform/sound_device.c \
+      --replace-fail "#include \"SDL_mixer.h\"" "#include \"SDL2/SDL_mixer.h\""
```
Comment (SuperSandro2000) [inline pkgs/games/julius/default.nix:36]: Can we add the correct ifdefs here and make that unconditional?

File: pkgs/games/julius/default.nix [hunk 2]
```diff
@@ -5,8 +5,15 @@
 , SDL2_mixer
 , cmake
 , libpng
+, runCommand
+, darwin ? null
```
Comment (SuperSandro2000) [inline pkgs/games/julius/default.nix:9]:
```suggestion
, darwin
```
that will never be null, even on linux

File: pkgs/games/julius/default.nix [hunk 3]
```diff
@@ -5,8 +5,15 @@
...
 }:
-
+let
+  darwinSymlinks = runCommand "julius-darwin-build-symlinks" { } ''
+    mkdir -p $out/bin
+    ln -s /usr/bin/sips /usr/bin/iconutil /usr/bin/codesign $out/bin
```
Comment (SuperSandro2000) [inline pkgs/games/julius/default.nix:14]: Wouldn't it be easier to append /usr/bin to PATH?

Comment (matteo-pacini) [reply pkgs/games/julius/default.nix:14]: Got inspired by a similar approach used in`macvim`: https://github.com/NixOS/nixpkgs/blob/nixos-24.05/pkgs/applications/editors/vim/macvim.nix#L18-L27

Happy to add it to the PATH if it's a better solution

File: pkgs/games/julius/default.nix (hunk 2 above)
Comment (matteo-pacini) [reply pkgs/games/julius/default.nix]: Good to know, thanks!

File: pkgs/games/julius/default.nix (hunk 1 above)
Comment (matteo-pacini) [reply pkgs/games/julius/default.nix]: Added this as part of the patch itself, using ifdefs as suggested

File: pkgs/games/julius/default.nix [hunk 4]
```diff
@@ -18,16 +25,37 @@ stdenv.mkDerivation rec {
...
+  ] ++ lib.optionals stdenv.isDarwin [
+    darwinSymlinks
+  ];
+
+  buildInputs = [
+    SDL2
+    SDL2_mixer
+    libpng
+  ] ++ lib.optionals stdenv.isDarwin [
+    darwin.apple_sdk_11_0.frameworks.Cocoa
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/games/julius/default.nix:45]: or just ``apple_sdk_11_0.callPackage``

**Comment by matteo-pacini** [comment]
@paparodeo @SuperSandro2000 ready for re-review, thanks for the feedback so far!

File: pkgs/games/julius/darwin-fixes.patch [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+@@ -869,9 +867,4 @@ else()
+         install(FILES "res/julius_512.png" DESTINATION "share/icons/hicolor/512x512/apps" RENAME "com.github.bvschaik.julius.png")
+     endif()
+ 
+-    # Unit tests
+-    if(${TARGET_PLATFORM} STREQUAL "default")
+-        enable_testing()
+-        add_subdirectory(test)
+-    endif()
+ endif()
```
Comment (paparodeo) [inline pkgs/games/julius/darwin-fixes.patch:48]:
```suggestion
```
i removed this and build still works on x64 darwin.

File: pkgs/games/julius/darwin-fixes.patch [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+index 0423e58f..ab2aef39 100644
+--- a/src/platform/sound_device.c
++++ b/src/platform/sound_device.c
+@@ -6,7 +6,12 @@
+ #include "platform/vita/vita.h"
+ 
+ #include "SDL.h"
++
++#ifdef __APPLE__
++#include "SDL2/SDL_mixer.h"
```
Comment (paparodeo) [inline pkgs/games/julius/darwin-fixes.patch:59]: this seems like something is off in cmake/FindSDL2mixer.cmake tho this seems to work too

Comment (matteo-pacini) [reply pkgs/games/julius/darwin-fixes.patch:49]: I think it expects a framework (?)

Comment (paparodeo) [reply pkgs/games/julius/darwin-fixes.patch:49]: for some reason it is adding `Headers` to the path it located, which seems incorrect. this patch fixes that.
```diff
diff --git a/cmake/FindSDL2_mixer.cmake b/cmake/FindSDL2_mixer.cmake
index 29cc683..1fa28a1 100644
--- a/cmake/FindSDL2_mixer.cmake
+++ b/cmake/FindSDL2_mixer.cmake
@@ -93,7 +93,6 @@ ELSE()
         PATH_SUFFIXES include include/SDL2
         PATHS ${SDL2_SEARCH_PATHS}
       )
-      set(SDL2_MIXER_INCLUDE_DIR "${SDL2_MIXER_INCLUDE_DIR}/Headers")
     endif()
 
     if(NOT APPLE OR NOT EXISTS "${SDL2_MIXER_INCLUDE_DIR}/SDL_mixer.h")
```

File: pkgs/games/julius/darwin-fixes.patch [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+     set_source_files_properties(${PROJECT_SOURCE_DIR}/res/julius.icns PROPERTIES
+@@ -687,7 +685,7 @@ if(APPLE)
+     # setting variables that will populate Info.plist
+     set(MACOSX_BUNDLE_GUI_IDENTIFIER "com.github.bvschaik.julius")
+     set(MACOSX_BUNDLE_BUNDLE_NAME ${USER_FRIENDLY_NAME})
+-    set(CMAKE_OSX_DEPLOYMENT_TARGET "10.10" CACHE STRING "Minimum OS X deployment version" FORCE)
++    set(CMAKE_OSX_DEPLOYMENT_TARGET "11.00" CACHE STRING "Minimum OS X deployment version" FORCE)
+     set(MACOSX_BUNDLE_ICON_FILE "julius.icns")
+     set(MACOSX_BUNDLE_BUNDLE_VERSION
+         "${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}${VERSION_REVISION}")
```
Comment (paparodeo) [inline pkgs/games/julius/darwin-fixes.patch:38]:
```suggestion
```
Is this needed? this just sets the minimum deployment target.

Comment (paparodeo) [reply pkgs/games/julius/darwin-fixes.patch:38]: or commented out. i think nix uses `stdenv.hostPlatform.darwinMinVersion` which defaults to `10.12` for x64-darwin.

File: pkgs/games/julius/darwin-fixes.patch (hunk 6 above)
Comment (matteo-pacini) [reply pkgs/games/julius/darwin-fixes.patch:49]: Much better, thanks!

File: pkgs/games/julius/darwin-fixes.patch (hunk 7 above)
Comment (matteo-pacini) [reply pkgs/games/julius/darwin-fixes.patch]: @paparodeo I am bit puzzled about the override at top level  now- should I keep `11.0` even though `x86-64-darwin` have a lower min. deployment target?

Comment (paparodeo) [reply pkgs/games/julius/darwin-fixes.patch]: so i think you can use the newer SDK and still have a lower deployment target. most of the reason i've seen overrideSDK used is when the build breaks due to using features from the newer sdk, like alligned_alloc. however, this package seems to build fine without the top-level override so it does seem unnecessary.

so i guess remove overrideSDK in top-level and then comment out or remove the line `set(CMAKE_OSX_DEPLOYMENT_TARGET "11.00" CACHE STRING "Minimum OS X deployment version" FORCE)` in the cmake file and then should be all done.

Comment (matteo-pacini) [reply pkgs/games/julius/darwin-fixes.patch]: Done

**Review by paparodeo — APPROVED** [review]
LGTM

**Comment by matteo-pacini** [comment]
@SuperSandro2000 friendly ping to ask for merge, thank you 🙏🏻

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**